docs/DifferentiableFunctions.md

```diff
@@ -0,0 +1,454 @@
+    ...
+}
+```
+
+Functions can have multiple `@differentiable` attributes with differentiable
+`wrt` parameter lists. `@differentiable` protocol requirements If a protocol
+requirement is marked with `@differentiable`, all implementations of the
+requirement are required to specify the same attribute. This enables generic
+code using differentiation defined in terms of protocol requirements.
+
+Here is an example of a neural network `Layer` protocol that defines a
+`@differentiable` required method called `applied(to:)`. As shown, the
+`applied(to:)` method can be differentiated in a `Layer` protocol extension,
+even though it is not a concrete method.
+
+```swift
+import TensorFlow
+
+/// A neural network layer.
+protocol Layer: Differentiable {
+    /// The input type of the layer.
+    associatedtype Input: Differentiable
+    /// The output type of the layer.
+    associatedtype Output: Differentiable
+    /// Returns the output obtained from applying the layer to the given input.
+    @differentiable
+    func applied(to input: Input) -> Output
+}
+
+extension Layer {
+    /// Returns the inference output and the backpropagation function obtained from applying the
+    /// layer to the given input.
+    ///
+    /// - Parameter input: The input to the layer.
+    /// - Returns: A tuple containing the output and the backpropagation function. The
+    ///   backpropagation function (a.k.a. backpropagator) takes a direction vector and returns the
+    ///   gradients at the layer and at the input, respectively.
+    func appliedForBackpropagation(to input: Input)
+        -> (output: Output,
+            backpropagator: (_ direction: Output.CotangentVector)
+                -> (layerGradient: CotangentVector, inputGradient: Input.CotangentVector)) {
+        let (out, pullback) = valueWithPullback(at: input) { layer, input in
+            return layer.applied(to: input)
+        }
+        return (out, pullback)
+    }
+}
+
+// Example neural network layer.
+struct DenseLayer: Layer {
+    var weight: Tensor<Float>
+    var bias: Tensor<Float>
+
+    @differentiable
+    func applied(to input: Tensor<Float>) -> Tensor<Float> {
+        return matmul(input, weight) + bias
+    }
+}
+
+// Example usage of `appliedForBackpropagation(to:)`.
+let dense = DenseLayer(weight: [[1, 1], [1, 1]], bias: [1, 1])
+let input: Tensor<Float> = [[3, 3]]
+let seed: Tensor<Float> = [[1, 1]]
+
+let (output, backprop) = dense.appliedForBackpropagation(to: input)
+let (𝛁dense, 𝛁input) = backprop(seed)
+
+dump(𝛁dense)
+// ▿ DenseLayer.AllDifferentiableVariables
+//   - weight: [[3.0, 3.0], [3.0, 3.0]]
+//   - bias: [1.0, 1.0]
+print(𝛁input)
+// [[2.0, 2.0]]
+```
+
+## Providing a custom derivative
+
+Use the `@differentiating` attribute to mark a function as a custom derivative
+for another function. This is useful for registering derivatives for primitive
+operations.
```

Comment:
"primitive operations" -> "functions that cannot be differentiated by traversing the function body, typically a primitive math operator in a library"

Reply:
Changing to:
```
Use the `@differentiating` attribute to mark a function as a custom derivative
for another function. This is useful for registering derivatives for functions
that cannot be differentiated by traversing the function body, typically
primitive math library operators.
```